vmaas_sync/metrics_cyndi.go

```diff
@@ -0,0 +1,125 @@
+	"app/base/database"
+	"app/base/models"
+	"app/base/utils"
+	"fmt"
+	"github.com/jinzhu/gorm"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"time"
+)
+
+const (
+	allSystemCount        = "allSystem"
+	uniqueTagsCount       = "uniqueTags"
+	systemsSapSystemCount = "systemsSapSystem"
+	systemsWithTagsCount  = "systemsWithTag"
+)
+
+var (
+	cyndiSystemsCnt = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "patchman_engine",
+		Subsystem: "vmaas_sync",
+		Name:      "cyndi_systems",
+	}, []string{"type"})
+
+	cyndiTagsCnt = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "patchman_engine",
+		Subsystem: "vmaas_sync",
+		Name:      "cyndi_tags_count",
+	}, []string{"type"})
+
+	tagsPerSystem = prometheus.NewHistogram(prometheus.HistogramOpts{
+		Namespace: "patchman_engine",
+		Subsystem: "vmaas_sync",
+		Name:      "tags_per_system",
+	})
+)
+
+type InventoryHosts struct {
+	SystemsCount int
+	UniqueTags   int
+	SapCount     int
+	TagsCount    int
+}
+
+func updateCyndiData() {
+	stats, err := getCyndiData()
+	if err != nil {
+		utils.Log("err", err.Error()).Error("unable to update cyndi metrics")
+		stats = InventoryHosts{}
+	}
+	cyndiTagsCnt.WithLabelValues(allSystemCount).Set(float64(stats.SystemsCount))
+	cyndiTagsCnt.WithLabelValues(uniqueTagsCount).Set(float64(stats.UniqueTags))
+	cyndiTagsCnt.WithLabelValues(systemsSapSystemCount).Set(float64(stats.SapCount))
+	cyndiTagsCnt.WithLabelValues(systemsWithTagsCount).Set(float64(stats.TagsCount))
+}
+
+func getCyndiData() (stats InventoryHosts, err error) {
+	err = database.Db.Table("inventory.hosts").
+		Select("COUNT(*) as systems_count, (SELECT COUNT(*) FROM inventory.hosts WHERE" +
+			" system_profile -> 'sap_system' = 'true') as sap_count, (SELECT COUNT(*) " +
+			"FROM (SELECT tags, cnt FROM (SELECT tags, COUNT(tags) over (partition by " +
+			"tags) cnt FROM inventory.hosts) as x where cnt = 1) as x) as unique_tags " +
+			", (SELECT COUNT(tags) FROM inventory.hosts) as tags_count").
+		Find(&stats).Error
+	if err != nil {
+		return stats, errors.Wrap(err, "unable to get inventory hosts from db")
+	}
+
+	return stats, nil
+}
+
+func updateCyndiTags() {
+	var counts []struct {
+		JSONArrayLength int
+	}
+
+	if err := database.Db.Table("inventory.hosts").
+		Select("json_array_length(tags::json)").
+		Find(&counts).Error; err != nil {
+		utils.Log("err", err.Error()).Error("Could not update cyndi tags metrics")
+		return
+	}
+	for _, c := range counts {
+		tagsPerSystem.Observe(float64(c.JSONArrayLength))
+	}
+}
+
+func updateCyndiSystemMetrics() {
+	counts, err := getCyndiCounts(time.Now())
+	if err != nil {
+		utils.Log("err", err.Error()).Error("unable to update cyndi system metrics")
+		return
+	}
+
+	for labels, count := range counts {
+		cyndiSystemsCnt.WithLabelValues(labels).Set(float64(count))
+	}
+}
+
+func getCyndiCounts(refTime time.Time) (map[string]int, error) {
+	systemsQuery := database.Db.Model(&models.InventoryHosts{})
```

Comment:
As this the only place of `models.InventoryHosts` struct usage, I think you can use `Table("inventory.hosts")` instead and not to define new struct at all. It's not needed if you don't use database row to go object mapping here.

Reply:
How is this resolved?